java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java

```diff
@@ -241,21 +261,22 @@ private void onMetricsResponse(Message msg) {
             f.set(userMetrics);
         }
         
-        private <T> SettableFuture<T> getFuture(Message msg) {
+        private SettableFutureTracker getFuture(Message msg) {
             long id = msg.getSourceId();
-            @SuppressWarnings("unchecked")
-            SettableFutureTracker futureTracker = futures.remove(id);
-            if (futureTracker == null) {
-                log.error(String.format("Future for message id %s not found as potentially it was a duplicate message "
-                        + "or was timed out in Java layer.", id));
-                throw new RuntimeException("Future for message id " + id + " not found as potentially it was a "
-                        + "duplicate message or was timed out in Java layer.");
-            }
-            // This cannot be null since SettableFuture class doesn't allow null values on future object
-            SettableFuture<T> f = (SettableFuture<T>) futureTracker.getFuture();
-            oldestFutureTrackerHeap.remove(futureTracker);
-            return f;
+            return getFutureTracker(id);
+        }
+    }
+
+    private SettableFutureTracker getFutureTracker(long id) {
+        @SuppressWarnings("unchecked")
+        SettableFutureTracker futureTracker = futures.remove(id);
+        if (futureTracker == null) {
+            String message = "Future for message id " + id + " not found";
```

Comment:
Can we keep the old error message, as some customers may rely on the format?

Reply:
done